tensorflow/core/kernels/data/shuffle_dataset_op_test.cc

```diff
@@ -0,0 +1,665 @@
+      /*seed2*/ DatasetOpsTestBase::CreateTensor<int64>(TensorShape({}), {2}),
+      /*reshuffle_each_iteration*/ true,
+      /*expected_outputs*/ ConvertToTensorVec<int64>({}),
+      /*expected_output_dtypes*/ {DT_INT64},
+      /*expected_output_shapes*/ {PartialTensorShape({})},
+      /*expected_cardinality*/ 10,
+      /*breakpoints*/ {0, 1, 9}};
+}
+
+class ParameterizedShuffleDatasetOpTest
+    : public ShuffleDatasetOpTest,
+      public ::testing::WithParamInterface<TestCase> {};
+
+TEST_P(ParameterizedShuffleDatasetOpTest, GetNext) {
+  int thread_num = 2, cpu_num = 2;
+  TestCase test_case = GetParam();
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
+
+  std::unique_ptr<OpKernel> shuffle_dataset_kernel;
+  TF_ASSERT_OK(CreateShuffleDatasetOpKernel(
+      test_case.reshuffle_each_iteration, test_case.expected_output_dtypes,
+      test_case.expected_output_shapes, &shuffle_dataset_kernel));
+
+  DatasetBase* range_dataset;
+  TF_ASSERT_OK(CreateRangeDataset<int64>(
+      test_case.range_data_param.start, test_case.range_data_param.end,
+      test_case.range_data_param.step, "range", &range_dataset));
+  Tensor range_dataset_tensor(DT_VARIANT, TensorShape({}));
+  TF_ASSERT_OK(
+      StoreDatasetInVariantTensor(range_dataset, &range_dataset_tensor));
+  Tensor buffer_size = test_case.buffer_size;
+  Tensor seed = test_case.seed;
+  Tensor seed2 = test_case.seed2;
+  gtl::InlinedVector<TensorValue, 4> inputs(
+      {&range_dataset_tensor, &buffer_size, &seed, &seed2});
+
+  std::unique_ptr<OpKernelContext> shuffle_dataset_context;
+  TF_ASSERT_OK(CreateShuffleDatasetContext(shuffle_dataset_kernel.get(),
+                                           &inputs, &shuffle_dataset_context));
+  DatasetBase* shuffle_dataset;
+  TF_ASSERT_OK(CreateDataset(shuffle_dataset_kernel.get(),
+                             shuffle_dataset_context.get(), &shuffle_dataset));
+  core::ScopedUnref scoped_unref_shuffle_dataset(shuffle_dataset);
+
+  std::unique_ptr<IteratorContext> iterator_ctx;
+  TF_ASSERT_OK(
+      CreateIteratorContext(shuffle_dataset_context.get(), &iterator_ctx));
+  std::unique_ptr<IteratorBase> iterator;
+  TF_ASSERT_OK(
+      shuffle_dataset->MakeIterator(iterator_ctx.get(), "Iterator", &iterator));
+
+  bool end_of_sequence = false;
+  std::vector<Tensor> out_tensors;
+  while (!end_of_sequence) {
+    std::vector<Tensor> next;
+    TF_EXPECT_OK(
+        iterator->GetNext(iterator_ctx.get(), &next, &end_of_sequence));
+    out_tensors.insert(out_tensors.end(), next.begin(), next.end());
+  }
+
+  // When `buffer_size = 1`, the output sequence of `ShuffleDataset` will be in
+  // order, so we need to consider the element sequence when evaluating the
+  // result for this case.
+  bool expect_items_equal = test_case.buffer_size.flat<int64>()(0) > 1;
+  TF_EXPECT_OK(ExpectEqual(out_tensors, test_case.expected_outputs,
```

Comment:
The shuffle order should be fixed given the seeds so we should be able to check for equality. 

On a related note, the `ExpectEqual` method is misleading. If it is actually not testing for equality (`expect_items_equal == false`), then it is not clear what it is doing? My suggestion would be to get rid of the boolean argument and make sure that all vector comparisons test for equality.

Reply:
Yes, we can check the equality with considering the element order.

Do you mean to change the above code to be as below
```
TF_EXPECT_OK(ExpectEqual(out_tensors, test_case.expected_outputs,
                           /*expect_items_equal*/ false));
```

Or do you mean to change the API of the utility function [`ExpectEqual`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/dataset_test_base.h#L54-L59) to remove the `expect_items_equal` argument?

The utility function [`ExpectEqual`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/dataset_test_base.h#L54-L59) wants to work like the python utility function [`assertDatasetProduces`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/kernel_tests/test_base.py#L138). When `expect_items_equal  == false`, it will return true if the two vectors have the same elements and same order; When `expect_items_equal == true`, it will return true if the two vectors have same elements (no need to have same order). The argument name `expect_items_equal` is confusing. How about changing it to be `regardless_of_order`?

Reply:
Rename the argument to `compare_order`.

Reply:
fixed via https://github.com/tensorflow/tensorflow/pull/27969/commits/6238de4711c8a0e5d25394374a1a32e5f5a8755d